core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/recipe/RecipeEngine.java

```diff
@@ -91,44 +95,45 @@ public void uploadUpscaleRecipes(Stack stack, HostGroup hostGroup, Set<HostMetad
 
     public void executePreAmbariStartRecipes(Stack stack, Set<HostGroup> hostGroups) throws CloudbreakException {
         Orchestrator orchestrator = stack.getOrchestrator();
-        if (shouldExecuteRecipeOnStack(stack, hostGroups, orchestrator)) {
+        if (shouldExecuteRecipeOnStack(hostGroups, orchestrator, PRE_AMBARI_START)) {
             orchestratorRecipeExecutor.preAmbariStartRecipes(stack);
         }
     }
 
+    // note: executed when LDAP config is present, because later the LDAP sync is hooked for this salt state in the top.sls.
     public void executePostAmbariStartRecipes(Stack stack, Set<HostGroup> hostGroups) throws CloudbreakException {
         Orchestrator orchestrator = stack.getOrchestrator();
-        if (shouldExecuteRecipeOnStack(stack, hostGroups, orchestrator)) {
+        if ((stack.getCluster() != null && stack.getCluster().getLdapConfig() != null) || recipesFound(hostGroups, POST_AMBARI_START)
+                && recipesSupportedOnOrchestrator(orchestrator)) {
             orchestratorRecipeExecutor.postAmbariStartRecipes(stack);
         }
     }
 
-    private boolean shouldExecuteRecipeOnStack(Stack stack, Set<HostGroup> hostGroups, Orchestrator orchestrator) throws CloudbreakException {
-        return ((stack.getCluster() != null && stack.getCluster().getLdapConfig() != null) || recipesFound(hostGroups))
-                && recipesSupportedOnOrchestrator(orchestrator);
-    }
-
-    public void executePostInstall(Stack stack) throws CloudbreakException {
+    public void executePostInstallRecipes(Stack stack, Set<HostGroup> hostGroups) throws CloudbreakException {
         Orchestrator orchestrator = stack.getOrchestrator();
-        if (recipesSupportedOnOrchestrator(orchestrator)) {
```

Comment:
salt highstate was executed even when there were no recipe --> this is why now highstate is executed only once (and this is expected in int test)